### PR TITLE
Material model for melt migration in the lowermost mantle

### PR DESCRIPTION
**Pull request by jdannberg**
This pull request contributes the material model we used in 

Dannberg, J., Myhill, R., Gassmöller, R., & Cottaar, S. (2021). The morphology, evolution and seismic visibility of partial melt at the core–mantle boundary: implications for ULVZs. Geophysical Journal International, 227(2), 1028-1059.

The pull request includes

- the material model
- a modification of the latent heat heating model that allows it to let the latent heat be computed self-consistently from the material model
- several test files that make sure the model works correctly
- a low-resolution version of the input file we used in the publication (I am happy to write a cookbook about it, but thought this pull request is long enough as it is)

I know the material model is quite extensive (but @bobmyhill told me this was the simplest we could make it), so suggestions for refactoring are appreciated. 

* [x] I have followed the [instructions for indenting my code](../blob/main/CONTRIBUTING.md#making-aspect-better).

### For new features/models or changes of existing features:

* [x] I have tested my new feature locally to ensure it is correct.
* [x] I have [created a testcase](http://www.math.clemson.edu/~heister/manual.pdf#sec%3Awriting_tests) for the new feature/benchmark in the [tests/](../blob/main/tests/) directory.
* [ ] I have added a changelog entry in the [doc/modules/changes](../blob/main/doc/modules/changes) directory that will inform other users of my change.

### CODE REVIEW

File: include/aspect/material_model/melt_boukare.h [hunk 1]
```diff
@@ -0,0 +1,356 @@
...
+        const double melting_reference_pressure = 120.e9;
+
+        double Fe_mantle_melting_temperature;                   // reference melting temperature for Fe mantle endmember at the reference pressure
+        double Mg_mantle_melting_temperature;                   // reference melting temperature for Mg mantle endmember at the reference pressure
+
+        const double Fe_mantle_melting_entropy = 33.77;         // molar entropy change of melting in J/mol K
+        const double Mg_mantle_melting_entropy = 34.33;         // molar entropy change of melting in J/mol K
+
+        const double Fe_mantle_melting_volume = 1.51e-07;       // molar volume change of melting of solid Fe mantle endmember in m3/mol
+        const double Mg_mantle_melting_volume = 9.29e-08;       // molar volume change of melting volume of solid Mg mantle endmember in m3/mol
```
Comment (gassmoeller) [inline include/aspect/material_model/melt_boukare.h:198]: Put the comments on separate lines

Comment (gassmoeller) [inline include/aspect/material_model/melt_boukare.h:198]: Put the comments on separate lines

File: source/material_model/melt_boukare.cc [hunk 2]
```diff
@@ -0,0 +1,1252 @@
+/*
+  Copyright (C) 2015 - 2019 by the authors of the ASPECT code.
```
Comment (gassmoeller) [inline source/material_model/melt_boukare.cc:2]:
```suggestion
  Copyright (C) 2015 - 2022 by the authors of the ASPECT code.
```

Comment (gassmoeller) [inline source/material_model/melt_boukare.cc:2]:
```suggestion
  Copyright (C) 2015 - 2022 by the authors of the ASPECT code.
```

File: source/material_model/melt_boukare.cc [hunk 3]
```diff
@@ -0,0 +1,1252 @@
...
+*/
+
+
+#include <aspect/adiabatic_conditions/interface.h>
+#include <aspect/material_model/melt_boukare.h>
+#include <aspect/gravity_model/interface.h>
+#include <aspect/simulator.h>
+
+#include <deal.II/base/parameter_handler.h>
+#include <deal.II/numerics/fe_field_function.h>
```
Comment (gassmoeller) [inline source/material_model/melt_boukare.cc:28]: it does not look like we need this header.

Comment (gassmoeller) [inline source/material_model/melt_boukare.cc:28]: it does not look like we need this header.

File: source/material_model/melt_boukare.cc [hunk 4]
```diff
@@ -0,0 +1,1252 @@
...
+          tait_parameters_a[i] = (1. + bulk_modulus_pressure_derivatives[i]) /
+                                 (1. + bulk_modulus_pressure_derivatives[i] + reference_bulk_moduli[i] * bulk_modulus_second_pressure_derivatives[i]);
+          tait_parameters_b[i] = bulk_modulus_pressure_derivatives[i] / reference_bulk_moduli[i]
+                                 - bulk_modulus_second_pressure_derivatives[i] / (1. + bulk_modulus_pressure_derivatives[i]);
+          tait_parameters_c[i] = (1. + bulk_modulus_pressure_derivatives[i] + reference_bulk_moduli[i] * bulk_modulus_second_pressure_derivatives[i]) /
+                                 (std::pow(bulk_modulus_pressure_derivatives[i],2) + bulk_modulus_pressure_derivatives[i]
+                                  - reference_bulk_moduli[i] * bulk_modulus_second_pressure_derivatives[i]);
+        }
+
+      febdg_idx = distance(endmember_names.begin(), find(endmember_names.begin(), endmember_names.end(), "FeSiO3_bridgmanite"));
```
Comment (gassmoeller) [inline source/material_model/melt_boukare.cc:91]: replace with `std::distance` here and the lines below

Comment (gassmoeller) [inline source/material_model/melt_boukare.cc:91]: replace with `std::distance` here and the lines below

File: source/material_model/melt_boukare.cc [hunk 5]
```diff
@@ -0,0 +1,1252 @@
...
+      wus_idx = distance(endmember_names.begin(), find(endmember_names.begin(), endmember_names.end(), "FeO_periclase"));
+      per_idx = distance(endmember_names.begin(), find(endmember_names.begin(), endmember_names.end(), "MgO_periclase"));
+      femelt_idx = distance(endmember_names.begin(), find(endmember_names.begin(), endmember_names.end(), "FeO_melt"));
+      mgmelt_idx = distance(endmember_names.begin(), find(endmember_names.begin(), endmember_names.end(), "MgO_melt"));
+    }
+
+    template <int dim>
+    double
+    MeltBoukare<dim>::
+    reference_viscosity () const
```
Comment (gassmoeller) [inline source/material_model/melt_boukare.cc:102]: No longer needed

Comment (gassmoeller) [inline source/material_model/melt_boukare.cc:102]: No longer needed

File: source/material_model/melt_boukare.cc [hunk 6]
```diff
@@ -0,0 +1,1252 @@
...
+          properties.heat_capacities[i] = Cp_ref + in.temperature[q] * dSdT;
+        }
+    }
+
+
+    template <int dim>
+    double
+    MeltBoukare<dim>::
+    endmember_thermal_energy (const double temperature,
+                              const unsigned int i) const
```
Comment (gassmoeller) [inline source/material_model/melt_boukare.cc:210]: `i` is pretty ambiguous. Maybe `endmember_index`?

Comment (gassmoeller) [inline source/material_model/melt_boukare.cc:210]: `i` is pretty ambiguous. Maybe `endmember_index`?

File: source/material_model/melt_boukare.cc [hunk 7]
```diff
@@ -0,0 +1,1252 @@
...
+    template <int dim>
+    double
+    MeltBoukare<dim>::
+    endmember_molar_heat_capacity (const double temperature,
+                                   const unsigned int i) const
+    {
+      AssertThrow(temperature > 0.0,
+                  ExcMessage("The temperature has to be larger than 0!"));
+
+      const double relative_T = Einstein_temperatures[i] / temperature;
```
Comment (gassmoeller) [inline source/material_model/melt_boukare.cc:232]: could you add the same replacement to the function above? only to make them easier to compare.

Comment (gassmoeller) [inline source/material_model/melt_boukare.cc:232]: could you add the same replacement to the function above? only to make them easier to compare.

File: source/material_model/melt_boukare.cc (hunk 6 above)
Comment (gassmoeller) [reply source/material_model/melt_boukare.cc:210]: same in the functions below

File: source/material_model/melt_boukare.cc [hunk 8]
```diff
@@ -0,0 +1,1252 @@
...
+    template <int dim>
+    double
+    MeltBoukare<dim>::
+    endmember_enthalpy_thermal_addition (const double temperature,
+                                         const unsigned int i) const
+    {
+      const double addition = reference_specific_heats[i] * temperature
+                              + 0.5 * specific_heat_linear_coefficients[i] * std::pow(temperature, 2.)
+                              - specific_heat_second_coefficients[i] / temperature
+                              + 2. * specific_heat_third_coefficients[i] * sqrt(temperature)
```
Comment (gassmoeller) [inline source/material_model/melt_boukare.cc:263]:
```suggestion
                              + 2. * specific_heat_third_coefficients[i] * std::sqrt(temperature)
```

Comment (gassmoeller) [inline source/material_model/melt_boukare.cc:263]:
```suggestion
                              + 2. * specific_heat_third_coefficients[i] * std::sqrt(temperature)
```

File: source/material_model/melt_boukare.cc [hunk 9]
```diff
@@ -0,0 +1,1252 @@
...
+                              - (reference_specific_heats[i] * reference_temperature
+                                 + 0.5 * specific_heat_linear_coefficients[i] * std::pow(reference_temperature, 2.)
+                                 - specific_heat_second_coefficients[i] / reference_temperature
+                                 + 2.0 * specific_heat_third_coefficients[i] * sqrt(reference_temperature));
```
Comment (gassmoeller) [inline source/material_model/melt_boukare.cc:267]:
```suggestion
                                 + 2.0 * specific_heat_third_coefficients[i] * std::sqrt(reference_temperature));
```

Comment (gassmoeller) [inline source/material_model/melt_boukare.cc:267]:
```suggestion
                                 + 2.0 * specific_heat_third_coefficients[i] * std::sqrt(reference_temperature));
```

File: source/material_model/melt_boukare.cc [hunk 10]
```diff
@@ -0,0 +1,1252 @@
...
+
+      return addition;
+    }
+
+
```
Comment (gassmoeller) [inline source/material_model/melt_boukare.cc:272]: add another empty line (check all empty lines between functions

Comment (gassmoeller) [inline source/material_model/melt_boukare.cc:272]: add another empty line (check all empty lines between functions

File: source/material_model/melt_boukare.cc [hunk 11]
```diff
@@ -0,0 +1,1252 @@
...
+    template <int dim>
+    double
+    MeltBoukare<dim>::
+    endmember_entropy_thermal_addition (const double temperature,
+                                        const unsigned int i) const
+    {
+      const double addition = reference_specific_heats[i] * std::log(temperature)
+                              + specific_heat_linear_coefficients[i] * temperature
+                              - 0.5 * specific_heat_second_coefficients[i] / std::pow(temperature, 2.)
+                              - 2.0 * specific_heat_third_coefficients[i] / sqrt(temperature)
```
Comment (gassmoeller) [inline source/material_model/melt_boukare.cc:282]: please check all occurences of `sqrt` in the file if they need a change
```suggestion
                              - 2.0 * specific_heat_third_coefficients[i] / std::sqrt(temperature)
```

Comment (gassmoeller) [inline source/material_model/melt_boukare.cc:282]: please check all occurences of `sqrt` in the file if they need a change
```suggestion
                              - 2.0 * specific_heat_third_coefficients[i] / std::sqrt(temperature)
```

File: source/material_model/melt_boukare.cc [hunk 12]
```diff
@@ -0,0 +1,1252 @@
...
+
+    template <int dim>
+    void
+    MeltBoukare<dim>::
+    convert_composition_to_fraction_of_endmembers (const double temperature,
+                                                   const double molar_Fe_in_solid,
+                                                   const double molar_Fe_in_melt,
+                                                   const std::vector<double> &endmember_gibbs_energies,
+                                                   std::vector<double> &endmember_mole_fractions_per_phase,
+                                                   double &f_pv) const
```
Comment (gassmoeller) [inline source/material_model/melt_boukare.cc:300]: In the header the last parameter is called `molar_bridgmanite_in_solid`, which is easier to understand. Please unify.

Comment (gassmoeller) [inline source/material_model/melt_boukare.cc:300]: In the header the last parameter is called `molar_bridgmanite_in_solid`, which is easier to understand. Please unify.

File: source/material_model/melt_boukare.cc [hunk 13]
```diff
@@ -0,0 +1,1252 @@
...
+
+      const double num_to_sqrt = -4. * molar_fraction_FeO * (partition_coefficient - 1.) * partition_coefficient * molar_fraction_SiO2
+                                 + std::pow(1. + (molar_fraction_FeO + molar_fraction_SiO2) * (partition_coefficient - 1.0), 2.);
+
+      endmember_mole_fractions_per_phase[febdg_idx] = (-1. + molar_fraction_FeO - (molar_fraction_FeO * partition_coefficient) + molar_fraction_SiO2 - (molar_fraction_SiO2 * partition_coefficient) + std::sqrt(num_to_sqrt)) /
+                                                      (2. * molar_fraction_SiO2 * (1. - partition_coefficient));
+
+      endmember_mole_fractions_per_phase[wus_idx] = endmember_mole_fractions_per_phase[febdg_idx] / (((1. - endmember_mole_fractions_per_phase[febdg_idx]) * partition_coefficient) + endmember_mole_fractions_per_phase[febdg_idx]);
+
+      f_pv = molar_fraction_SiO2; // molar fraction of bridgmanite in the solid
```
Comment (gassmoeller) [inline source/material_model/melt_boukare.cc:324]: this can move up to line 308.

Comment (gassmoeller) [inline source/material_model/melt_boukare.cc:324]: this can move up to line 308.

**Review by gassmoeller — COMMENTED** [review]
First batch of comments. I got to everything but melt_boukare.cc, in which I stopped at `compute_melt_molar_fraction` note for myself where to pick up).

**Comment by jdannberg** [comment]
I addressed your comments (and rebased, to make it compile without the reference viscosity).

File: source/material_model/melt_boukare.cc [hunk 14]
```diff
@@ -0,0 +1,1260 @@
...
+                                                        endmember_mole_fractions_per_phase,
+                                                        bridgmanite_molar_fraction_in_solid);
+
+
+          // convert from melt molar fraction to porosity
+          double solid_molar_volume = bridgmanite_molar_fraction_in_solid * endmember_mole_fractions_per_phase[febdg_idx] * endmembers.volumes[febdg_idx]
+                                      + bridgmanite_molar_fraction_in_solid * endmember_mole_fractions_per_phase[mgbdg_idx] * endmembers.volumes[mgbdg_idx]
+                                      + (1. - bridgmanite_molar_fraction_in_solid) * endmember_mole_fractions_per_phase[per_idx] * endmembers.volumes[per_idx]
+                                      + (1. - bridgmanite_molar_fraction_in_solid) * endmember_mole_fractions_per_phase[wus_idx] * endmembers.volumes[wus_idx];
+          double melt_molar_volume = endmember_mole_fractions_per_phase[mgmelt_idx] * endmembers.volumes[mgmelt_idx]
```
Comment (gassmoeller) [inline source/material_model/melt_boukare.cc:553]:
```suggestion
          const double melt_molar_volume = endmember_mole_fractions_per_phase[mgmelt_idx] * endmembers.volumes[mgmelt_idx]
```

File: source/material_model/melt_boukare.cc [hunk 15]
```diff
@@ -0,0 +1,1260 @@
...
+      for (unsigned int q=0; q<in.position.size(); ++q)
+        {
+          if (in.temperature[q] == 0.0)
+            {
+              if (melt_out != nullptr)
+                {
+                  const unsigned int porosity_idx = this->introspection().compositional_index_for_name("porosity");
+                  for (unsigned int q=0; q<in.position.size(); ++q)
+                    {
+                      double porosity = std::max(in.composition[q][porosity_idx],0.0);
```
Comment (gassmoeller) [inline source/material_model/melt_boukare.cc:586]:
```suggestion
                      const double porosity = std::max(in.composition[q][porosity_idx],0.0);
```

File: source/material_model/melt_boukare.cc [hunk 16]
```diff
@@ -0,0 +1,1260 @@
...
+          fill_endmember_properties(in, q, endmembers);
+
+          // We need the compositions of all phases.
+          const double solid_composition = in.composition[q][Fe_solid_idx];
+          double melt_composition = 0.0, melt_molar_fraction = 0.0;
+          double bridgmanite_molar_fraction_in_solid;
+
+          if (this->include_melt_transport())
+            {
+              const unsigned int Fe_melt_idx = this->introspection().compositional_index_for_name("molar_Fe_in_melt");
```
Comment (gassmoeller) [inline source/material_model/melt_boukare.cc:625]: move out of the loop

File: source/material_model/melt_boukare.cc [hunk 17]
```diff
@@ -0,0 +1,1260 @@
...
+      for (unsigned int q=0; q<in.temperature.size(); ++q)
+        {
+          std::vector<double> endmember_mole_fractions_per_phase(n_endmembers);
+          std::vector<double> endmember_mole_fractions_in_composite(n_endmembers);
+
+          fill_endmember_properties(in, q, endmembers);
+
+          // We need the compositions of all phases.
+          const double solid_composition = in.composition[q][Fe_solid_idx];
+          double melt_composition = 0.0, melt_molar_fraction = 0.0;
```
Comment (gassmoeller) [inline source/material_model/melt_boukare.cc:620]: put on separate lines

File: source/material_model/melt_boukare.cc [hunk 18]
```diff
@@ -0,0 +1,1260 @@
...
+          convert_composition_to_fraction_of_endmembers(in.temperature[q],
+                                                        solid_composition,
+                                                        melt_composition,
+                                                        endmembers.gibbs_energies,
+                                                        endmember_mole_fractions_per_phase,
+                                                        bridgmanite_molar_fraction_in_solid);
+
+          if (this->include_melt_transport())
+            {
+              const unsigned int porosity_idx = this->introspection().compositional_index_for_name("porosity");
```
Comment (gassmoeller) [inline source/material_model/melt_boukare.cc:638]: move out of loop. then you can also remove line 583

File: source/material_model/melt_boukare.cc [hunk 19]
```diff
@@ -0,0 +1,1260 @@
...
+          // Compute endmember molar fractions in the composite.
+          std::vector<double> phase_mole_fractions_in_composite(n_endmembers);
+          double solid_molar_mass = 0.0;
+          double melt_molar_mass = 0.0;
+          double melt_molar_volume = 0.0;
+          double solid_molar_volume = 0.0;
+
+          for (unsigned int i=0; i<n_endmembers; ++i)
+            {
+              // TODO: fix this using a phase vector (phase == 'bridgmanite')
```
Comment (gassmoeller) [inline source/material_model/melt_boukare.cc:657]: I guess this answers part of my comment above. But can you leave a better description than `fix this`? What is `this`?

File: source/material_model/melt_boukare.cc [hunk 20]
```diff
@@ -0,0 +1,1260 @@
...
+              if (endmember_states[i] == EndmemberState::solid)
+                {
+                  phase_mole_fractions_in_composite[i] = solid_molar_fraction * endmember_phase_fraction_in_solid;
+                  endmember_mole_fractions_in_composite[i] = phase_mole_fractions_in_composite[i] * endmember_mole_fractions_per_phase[i];
+                  solid_molar_mass += endmember_mole_fractions_in_composite[i] * molar_masses[i];
+                  solid_molar_volume += endmember_mole_fractions_in_composite[i] * endmembers.volumes[i];
+                }
+              else if (endmember_states[i] == EndmemberState::melt)
+                {
+                  phase_mole_fractions_in_composite[i] = melt_molar_fraction;
```
Comment (gassmoeller) [inline source/material_model/melt_boukare.cc:669]: this seems oddly unsymmetrical to line 662 above. Why do we multiply the `solid_molar_fraction` with something, but not the `melt_molar_fraction`?

File: source/material_model/melt_boukare.cc [hunk 21]
```diff
@@ -0,0 +1,1260 @@
...
+              for (unsigned int i=0; i<n_endmembers; ++i)
+                if (endmember_states[i] == EndmemberState::melt && melt_molar_volume > 0)
+                  melt_compressiblity += (endmember_mole_fractions_in_composite[i] * endmembers.volumes[i]) / (melt_molar_volume * endmembers.bulk_moduli[i]);
+
+              if (melt_molar_volume > 0)
+                melt_out->fluid_densities[q] = melt_molar_mass / melt_molar_volume;
+              else
+                {
+                  // make sure we have a useful melt density even if there is no melt to avoid density jumps
+                  double mass = 0, volume = 0;
```
Comment (gassmoeller) [inline source/material_model/melt_boukare.cc:716]: put on separate lines

File: source/material_model/melt_boukare.cc [hunk 22]
```diff
@@ -0,0 +1,1260 @@
...
+
+
+              if (endmember_states[i] == EndmemberState::solid && solid_molar_volume > 0)
+                {
+                  out.compressibilities[q] += (endmember_mole_fractions_in_composite[i] * endmembers.volumes[i])
+                                              / (solid_molar_volume * endmembers.bulk_moduli[i]);
+                }
+            }
+
+          if (solid_molar_volume > 0)
```
Comment (gassmoeller) [inline source/material_model/melt_boukare.cc:699]: Could you search through the file and replace all `0` that are used for doubles with `0.0`? It just makes the code harder to read if it is not immediately clear what type the variable has.

File: source/material_model/melt_boukare.cc [hunk 23]
```diff
@@ -0,0 +1,1260 @@
...
+                                                     * this->get_gravity_model().gravity_vector(in.position[q]);
+            }
+
+
+          if (this->include_melt_transport())
+            {
+              // Calculate the melting rate
+              const unsigned int Fe_solid_idx = this->introspection().compositional_index_for_name("molar_Fe_in_solid");
+              const unsigned int Fe_melt_idx = this->introspection().compositional_index_for_name("molar_Fe_in_melt");
+              const unsigned int porosity_idx = this->introspection().compositional_index_for_name("porosity");
```
Comment (gassmoeller) [inline source/material_model/melt_boukare.cc:738]: put these outside the loop at the top of the function (I think porosity_idx is already duplicated).

File: source/material_model/melt_boukare.cc [hunk 24]
```diff
@@ -0,0 +1,1260 @@
...
+                  boukare_out->bulk_composition[q] = bulk_composition;
+                  boukare_out->molar_volatiles_in_melt[q] = molar_volatiles_in_melt;
+                }
+
+              // We have to compute the update to the melt fraction in such a way that the bulk composition is conserved.
+              const double change_of_melt_composition = reaction_fraction * limit_update_to_0_and_1(old_melt_composition, melt_composition - old_melt_composition);
+              const double change_of_melt_fraction = reaction_fraction * limit_update_to_0_and_1(old_melt_molar_fraction, melt_molar_fraction - old_melt_molar_fraction);
+
+              melt_molar_fraction = old_melt_molar_fraction + change_of_melt_fraction;
+              melt_composition = old_melt_composition + change_of_melt_composition;
```
Comment (gassmoeller) [inline source/material_model/melt_boukare.cc:772]: reverse the order of these two lines to be consistent with the two lines directly above

File: source/material_model/melt_boukare.cc [hunk 25]
```diff
@@ -0,0 +1,1260 @@
...
+
+              if (melt_molar_fraction > 0 && melt_molar_fraction < 1)
```
Comment (gassmoeller) [inline source/material_model/melt_boukare.cc:774]: 0.0 and 1.0

File: source/material_model/melt_boukare.cc [hunk 26]
```diff
@@ -0,0 +1,1260 @@
...
+              convert_composition_to_fraction_of_endmembers(in.temperature[q],
+                                                            solid_composition,
+                                                            melt_composition,
+                                                            endmembers.gibbs_energies,
+                                                            endmember_mole_fractions_per_phase,
+                                                            bridgmanite_molar_fraction_in_solid);
+
+
+              // convert from melt molar fraction to porosity
+              double solid_molar_volume = bridgmanite_molar_fraction_in_solid * endmember_mole_fractions_per_phase[febdg_idx] * endmembers.volumes[febdg_idx]
```
Comment (gassmoeller) [inline source/material_model/melt_boukare.cc:790]: const

File: source/material_model/melt_boukare.cc [hunk 27]
```diff
@@ -0,0 +1,1260 @@
...
+                                          + bridgmanite_molar_fraction_in_solid * endmember_mole_fractions_per_phase[mgbdg_idx] * endmembers.volumes[mgbdg_idx]
+                                          + (1. - bridgmanite_molar_fraction_in_solid) * endmember_mole_fractions_per_phase[per_idx] * endmembers.volumes[per_idx]
+                                          + (1. - bridgmanite_molar_fraction_in_solid) * endmember_mole_fractions_per_phase[wus_idx] * endmembers.volumes[wus_idx];
+              double melt_molar_volume = endmember_mole_fractions_per_phase[mgmelt_idx] * endmembers.volumes[mgmelt_idx]
```
Comment (gassmoeller) [inline source/material_model/melt_boukare.cc:794]: const

File: source/material_model/melt_boukare.cc [hunk 28]
```diff
@@ -0,0 +1,1260 @@
...
+                                                       + (this->get_adiabatic_conditions().pressure(in.position[q]) - melting_reference_pressure) * Mg_mantle_melting_volume;
+                  double enthalpy_of_fusion = Fe_enthalpy_of_fusion * bulk_composition + Mg_enthalpy_of_fusion * (1.0-bulk_composition);
+                  enthalpy_of_fusion /= melt_molar_mass;
+
+                  enthalpy_out->enthalpies_of_fusion[q] = enthalpy_of_fusion;
+                }
+
+              // cutoff for viscosity at 30%
+              const double porosity = std::min(0.3, std::max(in.composition[q][porosity_idx],0.0));
+              out.viscosities[q] = (1.0 - porosity) * eta_0 * exp(- alpha_phi * porosity);
```
Comment (gassmoeller) [inline source/material_model/melt_boukare.cc:837]:
```suggestion
              out.viscosities[q] = (1.0 - porosity) * eta_0 * std::exp(- alpha_phi * porosity);
```

File: source/material_model/melt_boukare.cc [hunk 29]
```diff
@@ -0,0 +1,1260 @@
...
+          out.entropy_derivative_pressure[q]    = 0.0;
+          out.entropy_derivative_temperature[q] = 0.0;
+          out.thermal_conductivities[q] = thermal_conductivity;
+
+          for (unsigned int c=0; c<in.composition[q].size(); ++c)
+            out.reaction_terms[q][c] = 0.0;
+
+          double visc_temperature_dependence = 1.0;
+          const double delta_temp = in.temperature[q]-this->get_adiabatic_conditions().temperature(in.position[q]);
+          visc_temperature_dependence = std::max(std::min(std::exp(-thermal_viscosity_exponent*delta_temp/this->get_adiabatic_conditions().temperature(in.position[q])),1e8),1e-8);
```
Comment (gassmoeller) [inline source/material_model/melt_boukare.cc:860]: and delete line 858 above
```suggestion
          const double visc_temperature_dependence = std::max(std::min(std::exp(-thermal_viscosity_exponent*delta_temp/this->get_adiabatic_conditions().temperature(in.position[q])),1e8),1e-8);
```

File: source/material_model/melt_boukare.cc [hunk 30]
```diff
@@ -0,0 +1,1260 @@
...
+          out.viscosities[q] *= visc_temperature_dependence;
+        }
+
+      // fill melt outputs if they exist
+
```
Comment (gassmoeller) [inline source/material_model/melt_boukare.cc:865]:
```suggestion
```

File: source/material_model/melt_boukare.cc [hunk 31]
```diff
@@ -0,0 +1,1260 @@
...
+      if (melt_out != nullptr)
+        {
+          const unsigned int porosity_idx = this->introspection().compositional_index_for_name("porosity");
```
Comment (gassmoeller) [inline source/material_model/melt_boukare.cc:868]: move up and out

File: source/material_model/melt_boukare.cc [hunk 32]
```diff
@@ -0,0 +1,1260 @@
...
+              melt_out->permeabilities[q] = reference_permeability * std::pow(porosity,3) * std::pow(1.0-porosity,2);
+
+              // limit porosity to disaggregation threshold
+              porosity = std::min(0.3, porosity);
+
+              const double porosity_threshold = 0.01 * std::pow(this->get_melt_handler().melt_parameters.melt_scaling_factor_threshold, 1./3.);
+              melt_out->compaction_viscosities[q] = (1.0 - porosity) * xi_0 / std::max(porosity, porosity_threshold);
+
+              const double delta_temp = in.temperature[q]-this->get_adiabatic_conditions().temperature(in.position[q]);
+              const double visc_temperature_dependence = std::max(std::min(std::exp(-thermal_bulk_viscosity_exponent*delta_temp/this->get_adiabatic_conditions().temperature(in.position[q])),1e4),1e-4);
```
Comment (gassmoeller) [inline source/material_model/melt_boukare.cc:884]: so for the viscosity we use a bound of +/- 1e8, but for the compaction viscosity we use +/- 1e4. Maybe we can make this more clear, by extracting `const double viscosity_bound = 1e8;` and then doing `std::max(std::min(X,viscosity_bound),-viscosity_bound);` and the same for compaction viscosity?

File: source/material_model/melt_boukare.cc [hunk 33]
```diff
@@ -0,0 +1,1260 @@
...
+                             Patterns::List(Patterns::MultipleSelection("MgSiO3_bridgmanite|FeSiO3_bridgmanite|MgO_periclase|FeO_periclase|MgO_melt|FeO_melt|SiO2_melt")),
+                             "Names of the endmember components used in the equation of state and the melting model, "
+                             "and whose parameters are determined by the other input parameters of this material model. "
+                             "The order the parameters are given in has to be the same as the order the endmember names "
+                             "are given in. "
+                             "Units: none.");
+          prm.declare_entry ("Endmember states", "solid, solid, solid, solid, melt, melt, melt",
+                             Patterns::List(Patterns::MultipleSelection("solid|melt")),
+                             "States of the endmember components used in the equation of state and the melting model. "
+                             "For each endmember, this leist has to define if they belong to the melt or to the solid. "
```
Comment (gassmoeller) [inline source/material_model/melt_boukare.cc:1005]:
```suggestion
                             "For each endmember, this list has to define if they belong to the melt or to the solid. "
```

File: source/material_model/melt_boukare.cc [hunk 34]
```diff
@@ -0,0 +1,1260 @@
...
+                                                                                      n_endmembers,
+                                                                                      "Linear coefficients for specific heat polynomial");
+          specific_heat_second_coefficients = Utilities::possibly_extend_from_1_to_N (Utilities::string_to_double(Utilities::split_string_list(prm.get("Second coefficients for specific heat polynomial"))),
+                                                                                      n_endmembers,
+                                                                                      "Second coefficients for specific heat polynomial");
+          specific_heat_third_coefficients = Utilities::possibly_extend_from_1_to_N (Utilities::string_to_double(Utilities::split_string_list(prm.get("Third coefficients for specific heat polynomial"))),
+                                                                                     n_endmembers,
+                                                                                     "Third coefficients for specific heat polynomial");
+
+          //Check all lists have the correct length
```
Comment (gassmoeller) [inline source/material_model/melt_boukare.cc:1197]:
```suggestion
          // Check all lists have the correct length
```

File: source/material_model/melt_boukare.cc [hunk 35]
```diff
@@ -0,0 +1,1260 @@
...
+namespace aspect
+{
+  namespace MaterialModel
+  {
+    ASPECT_REGISTER_MATERIAL_MODEL(MeltBoukare,
+                                   "melt boukare",
+                                   "A material model that implements a melting model following Boukare"
+                                   "et al and uses it to compute the "
+                                   "material parameters required for the modelling of melt transport, "
+                                   "including a source term for the porosity according to a simplified.")
```
Comment (gassmoeller) [inline source/material_model/melt_boukare.cc:1258]: simplified ... . Also maybe include a note to look up the details in our paper. Just take the text from the header file.

**Review by gassmoeller — COMMENTED** [review]
Alright, only 35 comments left ;-).

File: source/material_model/melt_boukare.cc [hunk 36]
```diff
@@ -0,0 +1,1260 @@
...
+
+        const double p_Fe_mantle = std::exp(dG_Fe_mantle/(Fe_number_of_moles*R*T));
+        const double p_Mg_mantle = std::exp(dG_Mg_mantle/(Mg_number_of_moles*R*T));
+
+        // Mole composition of the solid and liquid (corresponds to the molar fraction of X_Fe, the iron-bearing endmember).
+        // In addition to the Phipps Morgan model, we also include volatiles here.
+        const double a = (p_Fe_mantle - p_Mg_mantle) * p_Fe_mantle/p_Mg_mantle;
+        const double b = p_Fe_mantle * (1. - 1./p_Mg_mantle) - molar_composition_of_bulk * (p_Fe_mantle - p_Mg_mantle)/p_Mg_mantle + molar_volatiles_in_bulk * (1. - p_Fe_mantle);
+        const double c = -molar_composition_of_bulk * (1. - 1./p_Mg_mantle);
+        double Xls = (-b + std::sqrt(b*b - 4.*a*c))/(2.*a);
```
[the remark this answers is not in the document]
Comment (jdannberg) [reply source/material_model/melt_boukare.cc:437]: I changed all the variable names in that function to reflect Appendix B in our paper.

**Comment by jdannberg** [comment]
Thanks for the review. I addressed all your comments.

**Review by gassmoeller — APPROVED** [review]
Ready to go when the testers are done.

**Review by gassmoeller — APPROVED** [review]

**Pull request merged**